File: frontend/src/services/api/answer-eval.ts
/**
 * Answer-eval API — baseline-vs-pipeline evaluation runs.
 */
import { apiFetch } from './client';

// Types

export interface AnswerEvalRun {
  id: number;
  project_id: string;
  question_count: number;
  judge_backend: string | null;
  baseline_groundedness: number | null;
  baseline_sufficiency: number | null;
  baseline_quality: number | null;
  pipeline_groundedness: number | null;
  pipeline_sufficiency: number | null;
  pipeline_quality: number | null;
  delta_groundedness: number | null;
  delta_sufficiency: number | null;
  delta_quality: number | null;
  status: 'running' | 'complete' | 'failed';
  created_at: string;
  finished_at: string | null;
  /** Project name — joined on the backend when listing globally. */
  project_name?: string | null;
}

export interface AnswerEvalResult {
  id: number;
  run_id: number;
  question: string;
  arm: 'baseline' | 'pipeline';
  answer_text: string | null;
  groundedness: number | null;
  sufficiency: number | null;
  quality: number | null;
  judge_reason: string | null;
  tokens: number | null;
  cost_usd: number | null;
  created_at: string;
}

export interface AnswerEvalRunDetail extends AnswerEvalRun {
  results: AnswerEvalResult[];
}

export interface ListRunsResponse {
  runs: AnswerEvalRun[];
}

export interface StartRunResponse {
  run_id: number;
}

// API object

export const answerEvalApi = {
  /** List eval runs — optionally scoped to a project. */
  listRuns: (projectId?: string) => {
    const query = projectId ? `?project_id=${encodeURIComponent(projectId)}` : '';
    return apiFetch<ListRunsResponse>(`/admin/answer-eval/runs${query}`);
  },

  /** Get a single run with its per-question results. */
  getRun: (id: number) => apiFetch<AnswerEvalRunDetail>(`/admin/answer-eval/runs/${id}`),

  /** Start a new eval run (async — returns immediately with run_id). */
  startRun: (projectId: string, n?: number) =>
    apiFetch<StartRunResponse>('/admin/answer-eval/run', {
      method: 'POST',
      body: JSON.stringify({ project_id: projectId, ...(n !== undefined ? { n } : {}) }),
    }),
};

File: backend/app/db/schema/_answer_eval.py
"""DDL for answer eval tables (Agentic-RAG answers, migration 154).

``answer_eval_runs`` is one evaluation batch (baseline vs pipeline across N
questions). ``answer_eval_results`` holds per-question per-arm scores.
FK cascade: deleting a run removes its results.
"""

from __future__ import annotations


def create_answer_eval_tables(conn) -> None:
    conn.execute(
        """
        CREATE TABLE IF NOT EXISTS answer_eval_runs (
            id                     INTEGER PRIMARY KEY AUTOINCREMENT,
            project_id             TEXT    NOT NULL,
            question_count         INTEGER NOT NULL DEFAULT 0,
            judge_backend          TEXT,
            baseline_groundedness  REAL,
            baseline_sufficiency   REAL,
            baseline_quality       REAL,
            pipeline_groundedness  REAL,
            pipeline_sufficiency   REAL,
            pipeline_quality       REAL,
            delta_groundedness     REAL,
            delta_sufficiency      REAL,
            delta_quality          REAL,
            status                 TEXT    NOT NULL DEFAULT 'running'
                CHECK (status IN ('running','complete','failed')),
            created_at             TEXT    NOT NULL DEFAULT (datetime('now')),
            finished_at            TEXT
        )
        """
    )
    conn.execute(
        """
        CREATE TABLE IF NOT EXISTS answer_eval_results (
            id            INTEGER PRIMARY KEY AUTOINCREMENT,
            run_id        INTEGER NOT NULL,
            question      TEXT    NOT NULL,
            arm           TEXT    NOT NULL CHECK (arm IN ('baseline','pipeline')),
            answer_text   TEXT,
            groundedness  REAL,
            sufficiency   REAL,
            quality       REAL,
            judge_reason  TEXT,
            tokens        INTEGER,
            cost_usd      REAL,
            created_at    TEXT    NOT NULL DEFAULT (datetime('now')),
            FOREIGN KEY (run_id)
                REFERENCES answer_eval_runs(id) ON DELETE CASCADE
        )
        """
    )
